### PR TITLE
Whisper - decoder and encoder optimization

### PR DESCRIPTION
**Pull request by mbahnasTT**
### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [x] [(Single-card) Demo tests (with perf runs)](https://github.com/tenstorrent/tt-metal/actions/runs/19908569939)
- [x] [(Single-card) Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/runs/19900956411)
- [x] [(Blackhole) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/19913908518)

### CODE REVIEW

**Review by idjuricTT — APPROVED** [review]

**Comment by idjuricTT** [comment]
Some end to end performance numbers:

- 30s audio on metal main branch - 21.45s
- 30s on this branch - 5.23s 🚀 
- 30s for two chips (with 15s chunks) - 3.23s
- UF target for 30s 1 chip - 4.63s


Accuracy check is still running:

https://github.com/tenstorrent/tt-shield/actions/runs/19820981760

EDIT: Accuracy check is also fine!

**Comment by nsextonTT** [comment]
Related side note- Currently seeing ~60% performance improvement in whisper decode t/s in my nanobind branch over baseline, so optimizing for minimal python-back-and-forth should see something similar.

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 1]
```diff
@@ -66,17 +66,35 @@ def calculate_key_values(config, key_value_states, *, parameters):
     bsz, tgt_len, hidden_size = key_value_states.shape
     head_size = hidden_size // config.encoder_attention_heads
 
-    fused_kv = key_value_states @ parameters.key_value.weight + parameters.key_value.bias
+    core_grid = ttnn.CoreGrid(y=8, x=8)
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Can the core grid be left unspecified so that the Blackhole version of the model uses more cores?

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 2]
```diff
@@ -66,17 +66,35 @@ def calculate_key_values(config, key_value_states, *, parameters):
...
+        num_heads=config.encoder_attention_heads,
+        num_kv_heads=0,
+    )[0]
+    key_states = ttnn.permute(key_states, [0, 1, 3, 2])
+
+    value_states = ttnn.experimental.nlp_create_qkv_heads(
+        value_states,
+        memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        num_heads=config.encoder_attention_heads,
+        num_kv_heads=0,
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:99]: can num_kv_heads be left unspecified if it already defaults to standard multi head attn?

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 3]
```diff
@@ -165,9 +274,33 @@ def whisper_attention(
...
+            key_states,
+            value_states,
+            scaling,
+            attention_mask,
+            is_cross_attention=is_cross_attention,
+            is_decode=is_decode,
+        )
     else:
-        fused_qkv = hidden_states @ parameters.query_key_value.weight + parameters.query_key_value.bias  # 1, S, 3xHxd
+        # fused_qkv = hidden_states @ parameters.query_key_value.weight + parameters.query_key_value.bias  # 1, S, 3xHxd
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: please delete comment

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 4]
```diff
@@ -165,9 +274,33 @@ def whisper_attention(
...
+        core_grid = ttnn.CoreGrid(y=8, x=8)
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: same comment as before about not hardcoding core grid if possible

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 5]
```diff
@@ -165,9 +274,33 @@ def whisper_attention(
...
+        if not is_decode:
+            fused_qkv_dtype = ttnn.bfloat8_b
+        else:
+            fused_qkv_dtype = ttnn.bfloat16
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:296]: does bfp8 yield insufficient acc for encoder?

File: models/demos/whisper/demo/demo.py [hunk 6]
```diff
@@ -723,6 +723,9 @@ def test_demo_for_conditional_generation(
         stream=stream,
     )
 
+    logger.info(f"TTFT: {ttft}")
+    logger.info(f"Decode throughput: {decode_throughput}")
+
```
Comment (skhorasganiTT) [inline models/demos/whisper/demo/demo.py]: TTFT and decode throughput are already logged in whisper_generator:_generate_with_temperature, please delete the extra logs.

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 7]
```diff
@@ -178,7 +311,7 @@ def whisper_attention(
             fused_qkv,
             num_heads=config.decoder_attention_heads,
             num_kv_heads=config.decoder_attention_heads,
-            transpose_k_heads=(not sdpa_with_kv_cache),
+            transpose_k_heads=no_transpose_k_heads,
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: transpose_k_heads=no_transpose_k_heads is misleading, can no_transpose_k_heads be renamed to transpose_k_heads?

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 8]
```diff
@@ -157,6 +258,14 @@ def whisper_attention(
 
     is_cross_attention = encoder_hidden_states is not None
     sdpa_with_kv_cache = not is_cross_attention and is_decode and kv_cache is not None
+    # Enabling encoder SDPA, to diable the K-transpose in ttnn.experimental.nlp_create_qkv_heads
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: diable -> disable

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 9]
```diff
@@ -125,18 +143,101 @@ def get_decode_sdpa_configs(config, device):
...
 
 
-def functional_sdpa(query_states, key_states, value_states, scaling, attention_mask):
-    query_states *= scaling
+def functional_sdpa(
+    query_states, key_states, value_states, scaling, attention_mask, is_cross_attention=False, is_decode=False
+):
+    if is_cross_attention and query_states.shape[1] == 20:
+        # num_heads must be 20 to be sharded
+        core_grid = ttnn.CoreGrid(y=4, x=5)
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:156]: So this will only work for whisper large v3 and distill large v3? Can you add a comment indicating that?

**Review by skhorasganiTT — COMMENTED** [review]
Thanks @mbahnasTT! Can you please ensure single card demos / frequent (filtered for whisper only) and blackhole demos still pass and post the links in the PR description?
- https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml
- https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml
- https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-demo-tests.yaml

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 1 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Removed core_grid

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 2 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:99]: This won't work since we are calling `nlp_create_qkv_heads` on K and V separately. So the input only contains 1 component, not 3 (QKV)

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 3 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Deleted

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 4 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Done

File: models/demos/whisper/demo/demo.py (hunk 6 above)
Comment (atupe-tt) [reply models/demos/whisper/demo/demo.py]: Removed log statements

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 7 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Renamed

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 8 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Corrected

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 9 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:156]: Added comment

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 5 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py:296]: Removed `fused_qkv_dtype = ttnn.bfloat8_b`

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py [hunk 10]
```diff
@@ -125,18 +154,102 @@ def get_decode_sdpa_configs(config, device):
...
+        k_chunk_size = 32
+        compute_grid_size = query_states.device().compute_with_storage_grid_size()
+        program_config = ttnn.SDPAProgramConfig(
+            compute_with_storage_grid_size=(compute_grid_size.x, compute_grid_size.y),
+            q_chunk_size=q_chunk_size,
+            k_chunk_size=k_chunk_size,
+            exp_approx_mode=True,  # NOTE: False is more correct
+        )
+
+        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
```
Comment (skhorasganiTT) [inline models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: I think this should be ttnn.init_device_compute_kernel_config with device.arch() passed in to generalize for wh/bh

**Review by skhorasganiTT — APPROVED** [review]

File: models/demos/whisper/tt/ttnn_optimized_functional_whisper.py (hunk 10 above)
Comment (atupe-tt) [reply models/demos/whisper/tt/ttnn_optimized_functional_whisper.py]: Fixed

**Comment by atupe-tt** [comment]
Reverted cross_attn cache implementation for now. Will be implemented in a different branch
